arch/x86/core/intel64/irq_offload.c

```diff
@@ -17,12 +17,32 @@
 extern void (*x86_irq_funcs[NR_IRQ_VECTORS])(const void *arg);
 extern const void *x86_irq_args[NR_IRQ_VECTORS];
 
+static void (*irq_offload_funcs[CONFIG_MP_NUM_CPUS])(const void *arg);
+static const void *irq_offload_args[CONFIG_MP_NUM_CPUS];
+
+void dispatcher(const void *arg)
+{
+	uint8_t cpu_id = _current_cpu->id;
+
+	if (irq_offload_funcs[cpu_id] != NULL) {
+		irq_offload_funcs[cpu_id](irq_offload_args[cpu_id]);
+	}
+}
 
 void arch_irq_offload(irq_offload_routine_t routine, const void *parameter)
 {
-	x86_irq_funcs[CONFIG_IRQ_OFFLOAD_VECTOR - IV_IRQS] = routine;
-	x86_irq_args[CONFIG_IRQ_OFFLOAD_VECTOR - IV_IRQS] = parameter;
+	int key = arch_irq_lock();
+	uint8_t cpu_id = _current_cpu->id;
+
+	x86_irq_funcs[CONFIG_IRQ_OFFLOAD_VECTOR - IV_IRQS] = dispatcher;
```

Comment:
FWIW this bit is now invariant and you could set it up once in an initialization hook somewhere, or use a static interrupt.